### ISSUE TITLE
kubectl: non-helpful error for 422 webhook status with no Reason field

### ISSUE THREAD

**Issue by nikhita** [kubernetes/kubernetes#111108]
### What happened?

kubectl apply displays different errors in v1.21 (and 1.22) vs v1.23 (and above) for the same webhook response, with the error message in v1.23 not being helpful at all.

For example, if a webhook returns a status with
- code `422` 
- `Reason` is not equal to `Invalid` / the `Reason` and `Details` fields are not populated

In v1.21.0, we see the following error. This is a helpful error message with plenty of context.

```
Error from server( could not find xxxx): error when creating "crd.yaml": admission webhook "xxx" denied the request: could not find kindXX "yy"
```

In v1.23.5, for the same webhook status, we'll see:

```
The request is invalid
```

This is not helpful at all.

### What did you expect to happen?

To get an equivalent helpful error message in v1.23.5

### How can we reproduce it (as minimally and precisely as possible)?

Make a webhook return a status with code `422` with `Reason` or `Details` not populated, or `Reason` not equivalent to `Invalid`.

### Anything else we need to know?

Looking at the code, it looks like this was a consequence of https://github.com/kubernetes/kubernetes/pull/101926. Since this PR was introduced in v1.23, this issue would exist for v1.22 as well.

kubectl first checks whether the error message is `IsInvalid`  (in both [1.21](https://github.com/kubernetes/kubernetes/blob/v1.21.0/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L145) and [1.23](https://github.com/kubernetes/kubernetes/blob/v1.23.5/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L148))
The definition of `IsInvalid` was changed in #101926.

**In 1.21** (and 1.22)
- `IsInvalid` just [checks if the](https://github.com/kubernetes/kubernetes/blob/v1.21.0/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L498-L500) reason [in the error is equivalent to the word](https://github.com/kubernetes/kubernetes/blob/v1.21.0/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L498-L500) "Invalid".
- Given that the webhook status does not include a `Reason`, the `StandardErrorMessage` [code block](https://github.com/kubernetes/kubernetes/blob/v1.23.5/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L184) is executed and kubectl shows a helpful error message.

**In 1.23** (and above)
- `IsInvalid` [checks](https://github.com/kubernetes/kubernetes/blob/v1.23.5/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L559-L568) if `reason == Invalid` OR if the status code is `422`, which is true for the webhook status.
- Since it’s true, this [codeblock](https://github.com/kubernetes/kubernetes/blob/v1.23.5/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L148-L163) is executed. `Details` is nil, so we just see the `The request is invalid` response. This error message is not helpful.

We should probably update how kubectl handles "invalid" errors and ensure that they display the `status.Message` field too (instead of just restricting to check the `Details`).

### Kubernetes version

Verified with v1.21.0 and v1.23.5. Have not verified this for v1.22, but it should also affect v1.22 due to the above reasons.

### Cloud provider

N/A

**Comment by akutz** [comment]
Hi @nikhita,

So this **purely** a client-side issue, right? The server is still sending back the exact same response/data structure? Thanks!

**Comment by akutz** [comment]
FYI, I posted [a question](https://kubernetes.slack.com/archives/C0EG7JC6T/p1661955619636249) in Kube slack related to this. We were hoping to alter our webhook behavior based on the originating user agent by passing the `*http.Request` into the Go context given to our webhook server's request handler. However, it appears we are not able to access the originating user-agent info. As @liggitt indicated in the Slack thread:

> altering the behavior of a webhook based on the originating client is a non-goal

In other words, the sooner _this_ issue is fixed, the better.

**Comment by liggitt** [comment]
checked with nikhita offline, will reroute the issue

/assign
/unassign @nikhita

**Comment by liggitt** [comment]
fix open in https://github.com/kubernetes/kubernetes/pull/112150
